client: bond reserves

chappjc opened the pull request:
<s>Requires https://github.com/decred/dcrdex/pull/2036. In draft until that's settled.</s>

This PR is the commits starting with `client/{asset,core}: RefundBond broadcasts too`.

Commit messages (but there are a ton of "fixup" commits already pushed that will be squashed before this is merged):

[client/{asset,core}: RefundBond broadcasts too](https://github.com/decred/dcrdex/commit/c759d01b4bb6e45a82c5f22bba779dd89d7b44b8) 

```
The asset.Bonder.RefundBond method now broadcasts, as it was supposed
to from the beginning. Only MakeBondTx required broadcasting in the
consumer.
```

[client/{asset,core,webserver}: BondLocked balance category](https://github.com/decred/dcrdex/commit/df1d6d4c06cba7a1b72dd056de7ed88225b2a963) 

```
This adds the client.WalletBalance.BondLocked field, which parallels
the ContractLocked field. These are used for core to supplement the
wallet's reported balance with values that the wallet does not
include in the balance itself because they are not spendable by the
wallet alone (core authors the spending tx).

This also updates the web UI's TypeScript to include bondLocked
when appropriate.  This also fixes a bug where contractLocked was
not included on the wallets page's "Locked" balance category.

This also adds to DCR's implementation fields for reserves tracking.
The values are included in the reported asset.Balance. Subsequent
commits will add the ability modify the reserves, and for funding
methods to respect the reserves.
```

[client/asset/dcr: respect reserves in funding](https://github.com/decred/dcrdex/commit/f9e6d9dc03755d4e79ba346ac49dd5f3268a809b) 

```
This updates the DCR wallet's transaction funding method and helpers to
respect the reserves added in the previous commit.

To do so without creating sizing transactions to actively isolate the
reserved value, we add new coin selection functions that we use exclude
UTXOs from transaction funding. The objective of these coin selection
functions is different from order funding coin selection.

The main helper, leastOverFund, attempts to pick a subset of the provided
UTXOs to reach the required amount with the objective of minimizing the
total amount of the selected UTXOs. This is different from the objective
used when funding orders, which is to minimize the number of UTXOs (to
minimize fees).

NOTE: The provided UTXO set MUST be sorted in ascending order (smallest
first, largest last)!

leastOverFund begins by partitioning the UTXO slice before the smallest
single UTXO that is large enough to fully fund the requested amount, if
it exists. If the smaller set is insufficient, the single largest UTXO
is returned. If instead the set of smaller UTXOs has enough total
value, it will search for a subset that reaches the amount with least
over-funding (see subsetSmallBias and subsetLargeBias). If that subset
has less combined value than the single sufficiently-large UTXO (if it
exists), the subset will be returned, otherwise the single UTXO will be
returned.

We also modify the "enough" functions and their constructors with
excess (change) reporting in a new return from the enough function.
The constructor also accepts a flag indicating if this returned value
should be set to zero. The tryFund helper now returns the final excess
value when it completes UTXO selection. This is used in the
(*ExchangeWallet).fund method, and in other callers, to discern if
there would be sufficient remaining in the wallet (including this
change) to satisfy a reserved amount. This amount is an input to fund
called "keep". If change should not be considered "kept" (e.g. no
preceding split txn, or mixing sends change to umixed account where it
is unusable for reserves), caller should return 0 extra from the enough
func (via the enough func's constructor flag).
```

[client/{asset,core}: reserves track as bonds are spend/refunded](https://github.com/decred/dcrdex/commit/4883a033d25454820f3d0506e25eee5eb9a976f5) 

```
This updates DCR's backend to begin modifying the reserves fields
as bonds are created (locked UTXOs) and spend (unlock UTXOs into
balance). Subsequent commits add the new exported wallet
interface methods for a consumer (core) to prepare and adjust the
reserves for the client's needs.

This also modifies the wallet's MakeBondTx method signature with
an additional return, an "abandon" function" to be used if the
consumer decides not to broadcast the generated bond transaction.
This function both unlocks the bond's funding utxos, and reverts
the reserves updates that were made when the transaction was
created.
```

[client/{asset,core}: ReserveBondFunds and dexAccount.tierChange](https://github.com/decred/dcrdex/commit/5c4029c396d10174d234d946c731fbe6b6c37d39) 

```
This adds the new Bonder methods, RegisterUnspent and ReserveBondFunds:

RegisterUnspent informs the wallet of a certain amount already locked in
unspent bonds that will eventually be refunded with RefundBond. This
should be used prior to ReserveBondFunds. This alone does not enable
reserves enforcement, and it should be called on bring-up when existing
bonds that may refund to this wallet are known. Once ReserveBondFunds is
called, even with 0 for future bonds, these live bond amounts will
become enforced reserves when they are refunded via RefundBond.

ReserveBondFunds (un)reserves funds for creation of future bonds.
MakeBondTx will create transactions that decrement these reserves, while
RefundBond will replenish the reserves. If the wallet's available
balance should be respected when adding reserves, the boolean argument
may be set to indicate this, in which case the return value indicates if
it was able to reserve the funds. In this manner, funds may be
pre-reserved so that when the wallet receives funds (from either
external deposits or refunding of live bonds), they will go directly
into locked balance. When the reserves are decremented to zero (by the
amount that they were incremented), all enforcement including any fee
buffering is disabled. Previously registered unspent/active bonds are
not forgotten however; they total amount of these are tracked in case
the wallet is used for reserves again in the future.

In core.Core, the new wallet reserves methods are used to maintain
adequate reserves based on the account's target tier and observed tier
change. The dexAccount type has new fields to support this:
- tierChange int64, is the total tier change that needs to be actuated
  with wallet reserves. Facilitates maintenance with changing tier.
- totalReserved int64, helps track and debug the total amount reserved
  with the wallet's ReserveBondFunds method *only* for this dexAccount.

Use of ReserveBondFunds and RegisterUnspent, facilitated by the new
dexAccount fields, happens in:
- connectWallets() via initialize(), after connectDEX(). Here it
  calls RegisterUnspent for all known bonds across all dexConnections.
- authDEX. On reconnect, any observed tier change is acctuated with
  wallet reserves. Initial auth on login is handled slightly
  differently, (re)reserving the full required amount given the known
  unspent bonds.
- handleTier change, which is called on penalization only, updates
  the tierChange field of the dexAccount.
- rotateBonds. Any tierChange value is actuated with reserves.
- PostBond. When making the initial bond for a new account, with
  maintenance enabled, ReserveBondFunds is used to pre-reseve. Here
  the boolean input is true, and the return is checked to ensure the
  wallet is sufficiently funded to increase reserves by the requested
  amount for the new bonds.
- UpdateBondOptions. When modifying target tier or changing the bond
  asset. The boolean input and output are also used here.
```

chappjc commented on client/core/bond.go:
```diff
@@ -672,54 +708,193 @@ func (c *Core) UpdateBondOptions(form *BondOptionsForm) error {
 		return err
 	}
+
+	bondAssets, _ := dc.bondAssets()
+	if bondAssets == nil {
+		return errors.New("dex not connected")
```
Just remembered I cannot do this otherwise user can't modify options (thus reserves) if the host is down.  Just need to update the two places below to do the nil check.

martonp commented on client/core/core.go:
```diff
@@ -2698,18 +2702,28 @@ func (dc *dexConnection) bondOpts() (assetID uint32, targetTier, max uint64) {
+}
+
+func (dc *dexConnection) hasActiveBond(assetID uint32) bool {
+	total, _ := dc.bondTotal(assetID)
```
This should check `active > 0`, not `total > 0`.

chappjc replied:
Hmm.  I guess I need two methods, or to rename this one to `hasUnspentBond`, because in the one current caller (`isActiveBondAsset`, just below), it's in the branch where: `includeLive should be set to consider all existing unspent bonds that need to be refunded in the future (only requires a broadcast, no wallet signing ability).` where the idea is you don't want to disable the wallet if you even need to refund an expired bond.

martonp replied:
Yeah, then rename would suffice.

chappjc commented:
Just a rebase, minimal review revisions

chappjc commented:
Squashed in fixups: https://github.com/decred/dcrdex/compare/0e01dc9460851fde6a51c8b93663cf9a5f1b040e..dd58c2eb3c9ab1ad147757d73f8134d2bd97de75

martonp reviewed (commented):
I think I found an issue while testing on simnet. I created a fresh account, posted 1 bonds worth, and at first the Locked amount showed 20+fees as I expected, but when I mined one extra block it jumped to 30 + fees.

<img width="363" alt="Screen Shot 2023-02-13 at 2 02 09 PM" src="https://user-images.githubusercontent.com/6186350/218550142-a02798b4-263c-4736-b3d0-32e85265601d.png">

chappjc commented:
> I think I found an issue while testing on simnet. I created a fresh account, posted 1 bonds worth, and at first the Locked amount showed 20+fees as I expected, but when I mined one extra block it jumped to 30 + fees.
> 
> <img alt="Screen Shot 2023-02-13 at 2 02 09 PM" width="363" src="https://user-images.githubusercontent.com/6186350/218550142-a02798b4-263c-4736-b3d0-32e85265601d.png">

OK, will have to go back to test fresh accounts.  I did some revisions to the system that I validated with existing accounts, but must have regressed with account creation.

chappjc commented on client/asset/dcr/coin_selection.go:
```diff
@@ -0,0 +1,210 @@
+	// Find a subset of the small UTXO set with smallest combined amount.
+	var set []*compositeUTXO
+	if sumUTXOs(small) >= amt {
+		set = subsetLargeBias(amt, small)
```
Was on my todo list.  :blush: Just threw together a few heiristics, but you have a point that it can be solved optimally.  Seems akin to the fractional knapsack problem (reversed).  Will give it a shot.

chappjc commented:
> > I think I found an issue while testing on simnet. I created a fresh account, posted 1 bonds worth, and at first the Locked amount showed 20+fees as I expected, but when I mined one extra block it jumped to 30 + fees.
> > <img alt="Screen Shot 2023-02-13 at 2 02 09 PM" width="363" src="https://user-images.githubusercontent.com/6186350/218550142-a02798b4-263c-4736-b3d0-32e85265601d.png">
> 
> OK, will have to go back to test fresh accounts. I did some revisions to the system that I validated with existing accounts, but must have regressed with account creation.

Oopsie fixed in https://github.com/decred/dcrdex/pull/2103/commits/638214db89050a6615c127a80ab96529c7fc29cb

buck54321 reviewed (commented):
I need to review **core** parts still, but I gotta run for a few hours at least. I think I get it all now. Thanks for all of the descriptive docs.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -1654,11 +1899,36 @@ func (dcr *ExchangeWallet) fund(enough func(sum uint64, size uint32, unspent *co
+		// unlike tryFund that minimizes the number of UTXOs chosen.
+		kept := leastOverFund(keep, utxos)
+		utxos = utxoSetDiff(utxos, kept)
+		sum, _, sz, coins, spents, redeemScripts, err = tryFund(utxos, enough)
```
If we get here, it is because the `enough` func is very likely not granting any spendable change i.e. no split tx.  Note the `if extra != 0 {` check above.  The logic in this commit is to try again but fist pick out utxos in the amount of at least `keep` before even calling `tryFund` -- that guarantees `keep` will be met but it can then fail if the pruned utxo set is insufficient for the order's `enough` func.

However (and I apologize for the moving target), after we chatted on matrix this morning, I switched the order of these attempts so that it _first_ sets aside utxos (using just the pruned set with `tryFund`), and if that fails to leave enough for to fund, it will then try with the full set and check if `extra` allowed it to satisfy `keep`: https://github.com/decred/dcrdex/pull/2103/commits/9f6ad97a87d0a29a7f9166efec3b37efa19cddc1

chappjc commented on client/asset/dcr/coin_selection.go:
```diff
@@ -0,0 +1,210 @@
+		// if sum >= amt { return utxos } // would have been i>=0 after break above
+		return nil
+	}
+	// if i == len(utxos)-1 { return utxos[i:] } // shouldn't happen if each in set are < amt
```
It's commented out here of course, but I'm not sure why I had written this as `utxos[i:]`, because that's the same as `utxos[i]` when `i == len(utxos)-1` (the last one).  As @martonp pointed out, it's not going to happen that just the one UTXO is sufficient because: `Each utxo in the input must be smaller than amt - use via leastOverFund only!` so this is commented.  Might as well remove.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -966,6 +1042,159 @@ func (dcr *ExchangeWallet) Balance() (*asset.Balance, error) {
+}
+
+func bondsFeeBuffer(highFeeRate uint64) uint64 {
+	const inputCount uint64 = 20 // plan for lots of inputs
```
Probably true in most cases, esp. with the high fee rate and the "tracks" multiplier below, but this helper is in `client/asset/dcr` so other assets would do differently.  I expect eth would be quite a bit simpler in this regard (no UTXO business to worry about), although this could be a lot of value for btc.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -1147,17 +1376,17 @@ func (dcr *ExchangeWallet) maxOrder(lotSize, feeSuggestion, maxFeeRate uint64) (
 	basicFee := dexdcr.InitTxSize * maxFeeRate
 	lots := avail / (lotSize + basicFee)
-	cfg := dcr.config()
+	// NOTE: Split tx is an order-time option. The max order is generally
+	// attainable when split is used, regardless of whether they choose it on
+	// the order form. Allow the split for max order purposes.
+	trySplitTx := true
```
Yeah, it's more nuanced than I let on, but this is referring to when there are reserves to respect, because the split maximizes the usable balance by allowing the split tx change to offset reserves.

However, the split actually eats a bit extra in fees because of the split tx baggage, but because `dcr.estimateSwap` falls back to skipping the split if needed, we can be sure that `maxOrder` calling `estimateSwap` with `trySplitTx=true` will give the highest estimate one way or another.  The comments in the revised tests in https://github.com/decred/dcrdex/pull/2103/commits/57db9ad43b8cb16a8bbc2fd0047c3bb2b2dd58d2 demonstrate this I think.

chappjc commented on client/core/core.go:
```diff
@@ -4803,7 +4836,7 @@ func (c *Core) MaxSell(host string, base, quote uint32) (*MaxOrderEstimate, erro
 
 // initializeDEXConnections connects to the DEX servers in the conns map and
-// authenticates the connection. This is done on Login.
+// authenticates the connection. This should be done only ONCE on Login.
```
Ah right, that's an atrocity.  Going to change that.  I'm also going to remove connectAccount now.

chappjc replied:
Oh heck, the `dc.acct.authed()` check in here makes this safe, so I'll remove the comment for now, but I figure we should factor out the loop contents for stuff like `AccountImport`.

buck54321 commented on client/core/bond.go:
```diff
@@ -424,6 +460,26 @@ func (c *Core) rotateBonds(ctx context.Context) {
+			}
+			c.log.Infof("Updating bond reserves by %s (automatic tier change adjustments)",
+				wallet.amtStringSigned(reserveDelta))
+			wallet.ReserveBondFunds(reserveDelta, false)
```
Is it possible for `rotateBonds` to run concurrently?

chappjc replied:
Should only be run from the loop in `watchBonds`.  Will document that it's not to be run willy nilly.

buck54321 reviewed (commented):
Just got through the **core** parts. A lot to digest, so I'm gonna swing back in a bit for a more thorough review. Just a couple of comments for now.

chappjc commented:
> Just got through the **core** parts. A lot to digest, so I'm gonna swing back in a bit for a more thorough review. Just a couple of comments for now.

Much appreciated.  It is a lot of work to digest.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -966,6 +1049,160 @@ func (dcr *ExchangeWallet) Balance() (*asset.Balance, error) {
+	// Normally we can plan on just 2 parallel "tracks" (single bond overlap
+	// when bonds are expired and waiting to refund) but that may increase
+	// temporarily if target tier is adjusted up.
+	const parallelTracks uint64 = 4
+	return parallelTracks * largeBondTxSize * highFeeRate
```
I need to really research this, but I observed them naturally merging sometimes.  Perhaps only likely with the high overlap ratio in simnet and testnet though.

chappjc commented:
Rebased and squashed away fixup commits, not retested yet, but there might be breakage with view-only merged.  Compared with original PR, there are three additional commits with qualitative changes beyond fixups to the other commits.  Minimal testing of the forced split with the extra output.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -1912,29 +1930,38 @@ func (dcr *ExchangeWallet) fund(keep uint64, // leave utxos for this reserve amt
+		// Without utxos set aside for keep, we have to consider any spendable
+		// change (extra) that the enough func grants us.
+		var extra uint64
+		sum, extra, sz, coins, spents, redeemScripts, err = tryFund(utxos, enough)
```
This commit also reverses the order of funding attempts vs the initial commit that updating `fund`/`tryFund`.  It tries straight away to set aside UTXOs for the reserves, and falls back to the standard funding approach, hoping that the enough func grants usable change in the extra output.

chappjc commented:
Tested forced split with extra output.  Worked as intended.

chappjc commented:
Just a rebase.  Nothing else revised.

chappjc commented on client/core/bond.go:
```diff
@@ -693,54 +730,200 @@ func (c *Core) UpdateBondOptions(form *BondOptionsForm) error {
+	if tierChanged {
+		dc.acct.targetTier = targetTier
+		acct.TargetTier = targetTier
+	}
```
It's permitted to change max bonded amt only.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -1060,6 +1060,141 @@ func (dcr *ExchangeWallet) Balance() (*asset.Balance, error) {
+	enforcedDelta := future + int64(feeBuffer)
+
+	// How much of that is covered by the available balance
+	if respectBalance {
```
I think `future > 0` would be the condition to indicate if reserves are increasing.

martonp reviewed (commented):
Working well while testing.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -1586,19 +1829,39 @@ func (dcr *ExchangeWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes
 
-	// Send a split, if preferred.
-	if useSplit && !ord.Immediate {
+	changeForReserves := useSplit && cfg.unmixedAccount == ""
```
Just the non-change outputs.  The change output goes into the unmixed account, according to signTxAndAddChange+depositAccount.
Mixing epochs are unfortunately kinda long so I don't think we can consider such change as available

martonp replied:
I see now.. but why does that change need to go into the unmixed account when we are doing a split? Couldn't it go right back into the trading account?

chappjc replied:
It's debatable, but all the input have been linked with each other.  They're also now linked with a dex order.

martonp replied:
Wouldn't the second output created for the reserves have the same issue then?

chappjc replied:
Yeah, and same as when an order that's funded by a split output is cancelled with no fill, that output stays in trading account.

But if we move nothing into unmixed, a wallet will quickly end up with nothing in the mixed account, I think.  Meaning the entire previously mixed set gets linked.  It would be down to deposits, redeems, and change on matched orders.  And the last category would not exist with splits in use.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -1842,11 +1845,25 @@ func (dcr *ExchangeWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes
+			useSplit = true
+			coins, redeemScripts, sum, inputsSize, err = dcr.fund(reserves,
+				orderEnough(ord.Value, ord.MaxSwapCount, bumpedMaxRate, true))
+			// And make an extra output for the reserves amount plus additional
```
Suggestion applied and squashed in.

chappjc commented on client/asset/interface.go:
```diff
@@ -473,14 +476,38 @@ type Broadcaster interface {
+	// the reserves are decremented to zero (by the amount that they were
+	// incremented), all enforcement including any fee buffering is disabled.
+	ReserveBondFunds(future int64, respectBalance bool) bool
+
```
We have that unused `asset.Balance.Other` field.

I expect frontend can work with the following or similar?

```diff
@@ -1061,7 +1062,9 @@ func (dcr *ExchangeWallet) Balance() (*asset.Balance, error) {
                dcr.log.Warnf("Available balance is below configured reserves: %f < %f",
                        toDCR(bal.Available), toDCR(reserves))
                reserves = bal.Available
+               bal.Other["Reserves Deficit"] = reserves - bal.Available
        }
+       bal.Other["Reserves"] = reserves
        bal.Available -= reserves
        bal.Locked += reserves
```

Or keys can be prefixed like `bal.Other["warning:Reserves Deficit"] = reserves - bal.Available` so it can be shown in red or something?

chappjc replied:
I this could work out

![image](https://user-images.githubusercontent.com/9373513/220485924-8bde0950-2022-4b0b-ab19-10b3c68e3d38.png)

I'd update that with `Bonds (locked)` too for completeness

chappjc replied:
I kinda like this, in the last commit

![image](https://user-images.githubusercontent.com/9373513/220500955-be813c8b-054e-48ce-b0f0-12cd270f4108.png)

martonp replied:
Maybe `Bonded` and `Bond Reserves`. Someone might not know what the reserves are for.

chappjc commented:
Ugh, the browser goes back to /register after restarting even if you've been bonded.  Something with view-only logic.

chappjc commented:
Oh nevermind I just deleted my dexc.db while it was running!  bad chappjc

chappjc commented:
I'm done with reviews on this, I think.

Only thing I have to check is what Wisdom discovered with legacy accounts, which is by no means a showstopper, but should be fixed:

> an account that paid legacy fee would have maxBondedAmt = 0
> if the user enables bond maintenance without setting a maxBondedAmt, the max value remains 0
> is this accounted for? so that the default max bonded amt is used?

martonp commented on client/asset/dcr/dcr.go:
```diff
@@ -4495,6 +4874,10 @@ func toDCR(v uint64) float64 {
 }
 
+func toDCRSigned(v int64) float64 {
+	return dcrutil.Amount(v).ToCoin()
```
This is exactly the same as `toDCR`.

chappjc commented:
> Only thing I have to check is what Wisdom discovered with legacy accounts, which is by no means a showstopper, but should be fixed:
> 
> > an account that paid legacy fee would have maxBondedAmt = 0
> > if the user enables bond maintenance without setting a maxBondedAmt, the max value remains 0
> > is this accounted for? so that the default max bonded amt is used?

Seems the missing condition was this https://github.com/decred/dcrdex/pull/2103/commits/479667e249540da5a66cc6e81440d161f963f079

Will test it out though.

JoeGruffins reviewed (approved):
Tested quite a bit on testnet and simnet. Looks fine. Only problem was it seems max send and max orders don't always work.

chappjc commented on client/asset/dcr/dcr.go:
```diff
@@ -1585,19 +1834,39 @@ func (dcr *ExchangeWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes
+	coins, redeemScripts, sum, inputsSize, err := dcr.fund(reserves,
+		orderEnough(ord.Value, ord.MaxSwapCount, bumpedMaxRate, changeForReserves))
+	if err != nil {
+		if !useSplit && reserves > 0 {
```
A big part of this forced split is that is does work for mixing wallets. That's why this is `!useSplit` instead of `!changeForReserves`.

martonp replied:
But the extra output goes to the trading wallet instead of the unmixed wallet.

chappjc replied:
That's fine, the trading account is used to fund stuff and counts toward available balance.  I might be missing your point about this?

martonp replied:
I don't understand why the forced split doesn't work for mixed wallets. Btw, if someone has a mixed wallet and `useSplit` is true, then the force will not be attempted, but if `useSplit` is false, then the force split will be attempted.

chappjc replied:
> I don't understand why the forced split doesn't work for mixed wallets.

It's certainly supposed to.  Do you mean you're trying it and it's not, or the intent of the code looks to you like it's supposed to block the force split?

> Btw, if someone has a mixed wallet and useSplit is true, then the force will not be attempted, but if useSplit is false, then the force split will be attempted.

That's a good point.  For regular non-mixing wallets that flow makes sense because regular splits have usable change, but not with mixing wallets, so a force split with the extra output should be attempted.

chappjc replied:
Ok, I see now what you mean that when useSplit is true and the wallet is mixed, it can't do the forced split.  Fixing.

ukane-philemon commented on client/core/bond.go:
```diff
@@ -680,54 +735,200 @@ func (c *Core) UpdateBondOptions(form *BondOptionsForm) error {
 	c.log.Debugf("Bond options for %v: target tier %d, bond asset %d, maxBonded %v",
 		acct.Host, dc.acct.targetTier, dc.acct.bondAsset, acct.MaxBondedAmt)
+
 	if err = c.db.UpdateAccountInfo(acct); err == nil {
 		success = true
-	}
+	} // else we might have already done ReserveBondFunds...
```
nit: we could just do
```
	err = c.db.UpdateAccountInfo(acct)
	success = err == nil // else we might have already done ReserveBondFunds...
```

chappjc commented:
Thanks to everyone who reviewed this PR.
